src/main/java/org/opensearch/flowframework/workflow/HttpHostStep.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.flowframework.workflow;
+
+import org.apache.hc.core5.http.HttpHost;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.support.PlainActionFuture;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.flowframework.exception.FlowFrameworkException;
+import org.opensearch.flowframework.util.ParseUtils;
+
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+
+import static org.opensearch.flowframework.common.CommonValue.HOSTNAME_FIELD;
+import static org.opensearch.flowframework.common.CommonValue.HTTP_HOST_FIELD;
+import static org.opensearch.flowframework.common.CommonValue.PORT_FIELD;
+import static org.opensearch.flowframework.common.CommonValue.SCHEME_FIELD;
+
+/**
+ * Step to register parameters for an HTTP Connection to a Host
+ */
+public class HttpHostStep implements WorkflowStep {
+
+    private static final Logger logger = LogManager.getLogger(HttpHostStep.class);
+    PlainActionFuture<WorkflowData> hostFuture = PlainActionFuture.newFuture();
+    static final String NAME = "http_host";
+
+    @Override
+    public PlainActionFuture<WorkflowData> execute(
+        String currentNodeId,
+        WorkflowData currentNodeInputs,
+        Map<String, WorkflowData> outputs,
+        Map<String, String> previousNodeInputs
+    ) {
+        Set<String> requiredKeys = Set.of(SCHEME_FIELD, HOSTNAME_FIELD, PORT_FIELD);
+        Set<String> optionalKeys = Collections.emptySet();
+
+        try {
+            Map<String, Object> inputs = ParseUtils.getInputsFromPreviousSteps(
+                requiredKeys,
+                optionalKeys,
+                currentNodeInputs,
+                outputs,
+                previousNodeInputs
+            );
+
+            String scheme = validScheme(inputs.get(SCHEME_FIELD));
+            String hostname = inputs.get(HOSTNAME_FIELD).toString();
+            int port = validPort(inputs.get(PORT_FIELD));
+
+            HttpHost httpHost = new HttpHost(scheme, hostname, port);
```

Comment:
Instead of HttpHost workflow step, what do you think about creating RestHightLevelClient workflow step? Anyway, we need to pass this step to `RestClientBuilder.builder()`. So why not do it in the same workflow step?

Reply:
> Instead of HttpHost workflow step, what do you think about creating RestHightLevelClient workflow step?

I have gone back and forth many times between these options, and written/deleted so much code that I think I can speak authoritatively here, and settled on separating it out this way.  Some reasons:
 - The client is a resource that needs to be closed, so is actually created in the same step as the REST API call.  Having a create client step introduces the complexity of closing it later.
 - Calling an external client requires this step; calling the cluster itself actually doesn't need an external configuration as it can be determined from the nodes themselves (which is how our integ tests do it).

i could wait and submit this all as part of a larger integrated setup so you could see why I chose it that way but I'm trying to submit incrementally here to make review easier.

Reply:
The above 3 reason makes sense. To understand the design from what you are thinking `HttpHost` would map to `ToolStep` and `RestClient` -> `AgentStep`, right?

What type of Clients would we have? Would we also have external clients wokflow step? If we will just have one client then having HttpHost in the ClientStep would be a better way to go. In short, all the depends on the number of client steps we will have.

The other concern I think, with the `RestAPIStep` and communicating plugins on the REST layer would introduce the same security issues we face for SDK. WDYT?

Reply:
> The above 3 reason makes sense. To understand the design from what you are thinking `HttpHost` would map to `ToolStep` and `RestClient` -> `AgentStep`, right?

Not quite.  
 - Yes, `HttpHostStep` is similar to `ToolStep` (in fact that's the class I copied and edited) in that it's a simple "configuration to create an object" class.  
   - It may be optional for http client calling the cluster.
 - There will be no "step" for the client. The actual client used is an implementation detail.
 - There will be a `RestApiStep` containing the details of an `HttpRequest` (method, endpoint, params, entity (body), and headers (which can include credentials).
   - I may later create a credentials step to separate those out, or maybe add them to the HttpHost step later.

> What type of Clients would we have? Would we also have external clients wokflow step? 

This is an implementation detail that the user doesn't need to know about. Options include:
   - HighLevelRestClient is the easiest option but is on path to deprecation
   - Apache has a RestClient
   - The OpenSearch Java Client is an option
   - Java has a built in HttpClient

It will be part of the RestAPI step as a closeable resource in a try-with-resources block.  Code will look something like this:
```java
    try (RestClient client = RestClient
            .builder(httpHost)
            .setHttpClientConfigCallback(httpClientBuilder -> {
                return httpClientBuilder
                        .setSSLContext(sslcontext)
                        .setDefaultCredentialsProvider(credentialsProvider);
            })
            // maybe other things here
            .build()) {
        // substitute the RestApiStep parameters in the Request here
        final Response response = client.performRequest(new Request("GET", "/_cluster/health"));
        // ... etc.
    }
```

> If we will just have one client then having HttpHost in the ClientStep would be a better way to go. In short, all the depends on the number of client steps we will have.

There will just be one RestApiStep with client entirely internal.  My thought was to reduce the duplication of those 3 elements.   Note that we could replace that host with the equivalent string: `https://hostname:port` and have that as the single line in the RestApiStep rather than a previous node input.  But adding credentials in the headers and picking the exact correct header name and doing all the appropriate obfuscating is something the user shouldn't have to do.

> The other concern I think, with the `RestAPIStep` and communicating plugins on the REST layer would introduce the same security issues we face for SDK. WDYT?

That's the way we eventually have to go for serverless and is an option to the existing Transport clients.  I don't think it's a concern at all for external connectors.  It's essentially already happening with some of our current steps:
 - Create connector step is similar to the http host step.... pointing to a particular API
 - Calling the predict API on a deployed model is similar to the rest api step, calling that external resource

There are definitely concerns calling the cluster over rest layer that we will need to be careful about.

Reply:
Just thinking aloud here. If the only inputs we are considering are the following three fields, what are your thoughts on using these fields as inputs for the `RestApiStep`? While having an `HttpHostStep` does aid in segregating logic, I don't have a strong opinion about keeping it as a separate step.
```
SCHEME_FIELD, HOSTNAME_FIELD, PORT_FIELD
```

Reply:
I see the following reasons you provided for having HttpHost as a separate step. I still think, it's fine to have this step under the same RestApiStep to reduce the workflow steps and the template size.

> The client is a resource that needs to be closed, so is actually created in the same step as the REST API call. Having a create client step introduces the complexity of closing it later.

We can have a special case of closing the client after the provisioning is done if the template includes RestApiStep.

> Calling an external client requires this step; calling the cluster itself actually doesn't need an external configuration as it can be determined from the nodes themselves (which is how our integ tests do it).

We can take the external client connection inputs as parameters to RestApiStep.

> Calling multiple APIs on an external client would have less repetition of these parameters than always including them in the rest API call itself.

Even if we have a single `HttpHostStep`, we might still have multiple REST API calls.

Required keys would include other credentials as well for the Rest call.
```java
Set<String> requiredKeys = Set.of(SCHEME_FIELD, HOSTNAME_FIELD, PORT_FIELD);
HttpHost httpHost = new HttpHost(scheme, hostname, port);
try (RestClient client = RestClient
            .builder(httpHost)
            .setHttpClientConfigCallback(httpClientBuilder -> {
                return httpClientBuilder
                        .setSSLContext(sslcontext)
                        .setDefaultCredentialsProvider(credentialsProvider);
            })
            // maybe other things here
            .build()) {
        // substitute the RestApiStep parameters in the Request here
        final Response response = client.performRequest(new Request("GET", "/_cluster/health"));
        // ... etc.
    }
```

Reply:
> Let me know WDYT and I will approve the PR based on that.

My intent with this PR was incremental changes to avoid a massive PR later.  I may change this later, and even consider deleting it if we find we don't need it.

As-is, the scheme/hostname/port can be replaced with a single string and parsed, so provide little value.  It's the future inclusion of credentials that I am preparing for.

My suggestion:
1. approve as-is knowing it will change later
2. close this PR and wait for the complete RestAPI implementation later, where its usage is a bit more obvious and can be more carefully evaluated.

I don't have a strong opinion either way, I'll keep this commit and iterate on it for the future RestAPI step

Reply:
>Calling an external client requires this step; calling the cluster itself actually doesn't need an external configuration as it can be determined from the nodes themselves (which is how our integ tests do it).

What do you mean by not needing a client to call the cluster itself (another plugin)?

>As-is, the scheme/hostname/port can be replaced with a single string and parsed, so provide little value. It's the future inclusion of credentials that I am preparing for.

if you are including the header in the `RestAPIStep` potentially, would you have the `httphost` stay mostly constant for all other calls inside the cluster? so we don't have to repeat this in every step basically?

Reply:
That sounds good to me, I'm okay keeping with this, I think my initial confusion came from the fact that I thought we were first gonna focus on opensearch APIs (considered everything not in plugin as external) and later on external APIs (outside of opensearch).